### PR TITLE
*: use SimSIMD for vectors

### PR DESCRIPTION
**Pull request by Lloyd-Pottiger**
### What problem does this PR solve?

Issue Number: ref #9032 

Problem Summary:

### What is changed and how it works?

Pick https://github.com/tidbcloud/tiflash-cse/pull/221, https://github.com/tidbcloud/tiflash-cse/pull/224, https://github.com/tidbcloud/tiflash-cse/pull/262 and https://github.com/tidbcloud/tiflash-cse/pull/245

```commit-message
*: use SimSIMD for vectors
```

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

**Comment by JaySon-Huang** [comment]
In order to resolve the compile error below, you need to pick these changes: https://github.com/tidbcloud/tiflash-cse/pull/262 @Lloyd-Pottiger 
```
ld.lld: error: undefined symbol: DB::DM::VectorIndexHNSWSIMDFeatures::get()
>>> referenced by TiFlashBuildInfo.cpp
>>>               TiFlashBuildInfo.cpp.o:(TiFlashBuildInfo::getEnabledFeatures()) in archive dbms/libtiflash_common_iod.a
ld.lld: error: undefined symbol: DB::VectorDistanceSIMDFeatures::get()
>>> referenced by TiFlashBuildInfo.cpp
>>>               TiFlashBuildInfo.cpp.o:(TiFlashBuildInfo::getEnabledFeatures()) in archive dbms/libtiflash_common_iod.a
```

File: .devcontainer/README.md
Comment (JaySon-Huang) [inline .devcontainer/README.md:1]: should revert these changes

File: release-centos7-llvm/env/README.md
Comment (JaySon-Huang) [inline release-centos7-llvm/env/README.md:1]: should revert these changes

File: dbms/src/TiDB/Decode/Vector.cpp [hunk 1]
```diff
@@ -50,62 +51,75 @@ Float64 VectorFloat32Ref::l2SquaredDistance(VectorFloat32Ref b) const
 {
     checkDims(b);
 
-    Float32 distance = 0.0;
-    Float32 diff;
-
-    for (size_t i = 0, i_max = size(); i < i_max; ++i)
+    static simsimd_metric_punned_t metric = nullptr;
+    if (metric == nullptr)
```
Comment (CalvinNeo) [inline dbms/src/TiDB/Decode/Vector.cpp:55]: Can `VectorFloat32Ref::l2SquaredDistance` be called concurrently?

File: dbms/src/VectorSearch/DistanceSIMDFeatures.cpp [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+
+namespace DB
+{
+
+std::vector<std::string> VectorDistanceSIMDFeatures::get()
+{
+    simsimd_capability_t cap_l2 = simsimd_details::actual_capability(simsimd_datatype_f32_k, simsimd_metric_l2sq_k);
+    simsimd_capability_t cap_cos = simsimd_details::actual_capability(simsimd_datatype_f32_k, simsimd_metric_cos_k);
+
+    auto cap_to_string = [](simsimd_capability_t cap) -> std::string {
```
Comment (CalvinNeo) [inline dbms/src/VectorSearch/DistanceSIMDFeatures.cpp:55]: Could there be `cap == simsimd_cap_sve_k & simsimd_cap_sve2_k`?

Comment (Lloyd-Pottiger) [reply dbms/src/VectorSearch/DistanceSIMDFeatures.cpp:55]: https://github.com/ashvardanian/SimSIMD/blob/96891ce929e3ec6b3600f12c82c383cf1d076548/include/simsimd/simsimd.h#L444

seems impossible.

File: dbms/src/TiDB/Decode/Vector.cpp (hunk 1 above)
Comment (Lloyd-Pottiger) [reply dbms/src/TiDB/Decode/Vector.cpp]: Good catch! Fixed in https://github.com/pingcap/tiflash/pull/9372/commits/b427bee1104dfaada45901856124f2344777fc79

File: dbms/src/VectorSearch/DistanceSIMDFeatures.cpp (hunk 2 above)
Comment (CalvinNeo) [reply dbms/src/VectorSearch/DistanceSIMDFeatures.cpp:55]: Then just use `==` rather than `&`?
If it could be possible in further versions, then move `sve2` in the front of `sve`?

File: libs/libdaemon/src/BaseDaemon.cpp [hunk 3]
```diff
@@ -1175,7 +1175,10 @@ void BaseDaemon::logVersion() const
     LOG_INFO(log, "Welcome to TiFlash");
     std::stringstream ss;
     TiFlashBuildInfo::outputDetail(ss);
-    LOG_INFO(log, "TiFlash build info: {}", ss.str());
+
+    std::string line;
+    while (std::getline(ss, line, '\n'))
+        LOG_INFO(log, "{}", line);
```
Comment (CalvinNeo) [inline libs/libdaemon/src/BaseDaemon.cpp:1181]: What is the difference?

File: dbms/src/VectorSearch/DistanceSIMDFeatures.cpp (hunk 2 above)
Comment (Lloyd-Pottiger) [reply dbms/src/VectorSearch/DistanceSIMDFeatures.cpp:55]: addressed

File: libs/libdaemon/src/BaseDaemon.cpp (hunk 3 above)
Comment (Lloyd-Pottiger) [reply libs/libdaemon/src/BaseDaemon.cpp:1181]: print multiple log

Comment (Lloyd-Pottiger) [reply libs/libdaemon/src/BaseDaemon.cpp]: revert

**Review by CalvinNeo — APPROVED** [review]
lgtm
